csi/controller_server.go

```diff
@@ -254,6 +254,18 @@ func (cs *ControllerServer) isNodeReady(nodeID string) bool {
 // ControllerPublishVolume will attach the volume to the specified node
 func (cs *ControllerServer) ControllerPublishVolume(ctx context.Context, req *csi.ControllerPublishVolumeRequest) (*csi.ControllerPublishVolumeResponse, error) {
 	logrus.Infof("ControllerServer ControllerPublishVolume req: %v", req)
+	if req.GetNodeId() == "" {
```

Comment:
Put this check in front of this part. Or directly update the Error Code of this error instead :
```
	if !cs.isNodeReady(req.GetNodeId()) {
		msg := fmt.Sprintf("ControllerPublishVolume: the volume %s cannot be attached to `NotReady` node %s",
			req.GetVolumeId(), req.GetNodeId())
		logrus.Warn(msg)
		return nil, status.Error(codes.NotFound, msg)
	}
```

Reply:
The issue with the approach to check for the node would fail other tests. 

To change the err code would fail this [example](https://github.com/kubernetes-csi/csi-test/blob/master/pkg/sanity/controller.go#L961) where is expecting `codes.NotFound`.

But without change the err code would fail other tests where expecting different error code. Most negative tests check the request with the minimal mocking of the request. For [example](https://github.com/kubernetes-csi/csi-test/blob/master/pkg/sanity/controller.go#L866), this one does not mock the NodeId. 

By putting the node check right to the front not only fail by the unexpected error code, and also bypasses the request that is testing for.

However, I could still move it forward, together with the other checks...

Reply:
OK, make sense.

Then moving this check together with `if !cs.isNodeReady(req.GetNodeId())` may be enough.

Reply:
I've moved this forward and it goes before the capability check. This is because the node id [test](https://github.com/kubernetes-csi/csi-test/blob/master/pkg/sanity/controller.go#L870-L873) doesn't mock capability in the request. Though they result in the same error code, they are actually in a certain sequence to ensure different tests go through the correct checks.

Reply:
All right, then the current implementation is good enough.